src/equations/shallow_water_2d.jl

```diff
@@ -0,0 +1,619 @@
+
+The bottom topography function ``b(x,y)`` is set inside the initial condition routine
+for a particular problem setup. To test the conservative form of the SWE one can set the bottom topography
+variable `b` to zero.
+
+In addition to the unknowns, Trixi currently stores the bottom topography values at the approximation points
+despite being fixed in time. This is done for convenience of computing the bottom topography gradients
+on the fly during the approximation as well as computing auxiliary quantities like the total water height ``H``
+or the entropy variables.
+This affects the implementation and use of these equations in various ways:
+* The flux values corresponding to the bottom topography must be zero.
+* The bottom topography values must be included when defining initial conditions, boundary conditions or
+  source terms.
+* [`AnalysisCallback`](@ref) analyzes this variable.
+* Trixi's visualization tools will visualize the bottom topography by default.
+
+References for the SWE are many but a good introduction is available in Chapter 13 of the book:
+- Randall J. LeVeque (2002)
+  Finite Volume Methods for Hyperbolic Problems
+  [DOI: 10.1017/CBO9780511791253](https://doi.org/10.1017/CBO9780511791253)
+"""
+struct ShallowWaterEquations2D{RealT<:Real} <: AbstractShallowWaterEquations{2, 4}
+  gravity::RealT # gravitational constant
+end
+
+function ShallowWaterEquations2D(gravity_constant)
+  ShallowWaterEquations2D(gravity_constant)
+end
+
+
+have_nonconservative_terms(::ShallowWaterEquations2D) = Val(true)
+varnames(::typeof(cons2cons), ::ShallowWaterEquations2D) = ("h", "h_v1", "h_v2", "b")
+varnames(::typeof(cons2prim), ::ShallowWaterEquations2D) = ("H", "v1", "v2", "b") # total water height: H = h + b
+
+
+# Set initial conditions at physical location `x` for time `t`
+"""
+    initial_condition_constant(x, t, equations::ShallowWaterEquations2D)
+
+A constant initial condition to test free-stream preservation / well-balancedness.
+"""
+function initial_condition_well_balancedness(x, t, element_id, equations::ShallowWaterEquations2D)
+  # Set the background values
+  H = 3.0
+  v1 = 0.0
+  v2 = 0.0
+  b = 0.0
+
+  # Setup a discontinuous bottom topography using the element id number
+  if element_id == 7
+    b = 2.0 + 0.5 * sin(2.0 * pi * x[1]) + 0.5 * cos(2.0 * pi * x[2])
+  end
+
+  return prim2cons(SVector(H, v1, v2, b), equations)
+end
+
+
+"""
+    initial_condition_convergence_test(x, t, equations::ShallowWaterEquations2D)
+
+A smooth initial condition used for convergence tests in combination with
+[`source_terms_convergence_test`](@ref)
+(and [`BoundaryConditionDirichlet(initial_condition_convergence_test)`](@ref) in non-periodic domains).
+"""
+function initial_condition_convergence_test(x, t, element_id, equations::ShallowWaterEquations2D)
+  # some constants are chosen such that the function is periodic on the domain [0,sqrt(2)]^2
+  c  = 7.0
+  ωx = 2.0 * pi * sqrt(2.0)
+  ωt = 2.0 * pi
```

Comment:
Please no non-ASCII code :grimacing:

Reply:
okay, I can switch this to be, e.g., `omega_x`. Out of curiosity is it the `x` appended to `ω` that causes the issue? I ask because characters like `ω` and `γ` are used in `source_terms_convergence_test` in the `compressible_euler_2d.jl`. I am just trying to keep track on when it is okay (if ever) to use such characters.

Reply:
I unresovled this to let @sloede answer the question here

Reply:
Well, there is
> Executable code should only use ASCII characters.

from the [style guide](https://trixi-framework.github.io/Trixi.jl/stable/styleguide/#Style-guide), so my comment was mainly aimed at the use of Greek letters. However, I *do* think that the underscore makes this more readable, but YMMV.

Regarding `source_terms_convergence_test`: This is a classic case of what I call the "constancy of bad decisions". The referenced function is probably the 3rd or 4th IC *ever* implemented in Trixi, and back than we were still in awe of Julia's shininess. Since it was me who set this horrible precedence :see_no_evil: , feel free to leave it as it is, since consistency is already broken for the convergence test functions anyways :shrug:

Reply:
I decided to remove the non-ASCII text